### PR TITLE
[Bugfix][Model] Fix DiffusionGemma silently freezing attention mask under CUDA graph replay

### PR DESCRIPTION
**Pull request by fjosw**
## Purpose

While looking into whether FULL CUDA graphs are worth the added complexity for DiffusionGemma, I found that `DiffusionGemmaModelState._causal_buf` (the per-request flag that switches each request between causal/encoder and bidirectional/denoise attention) stops being updated once `CUDAGraphMode.FULL` graphs are captured. It's bool, but FlashAttentionMetadataBuilder.build() casts non-int32 causal tensors via `causal.to(torch.int32)`, allocating a new tensor every call. A FULL-mode graph binds to the address it saw at capture time, so once captured it keeps reading that orphaned tensor forever. Later updates to `_causal_buf` never reach it. Any request served through an already-captured FULL graph gets whatever causal/bidirectional pattern existed at capture (warmup) time for its slot, for the graph's whole lifetime, silently.

This isn't opt-in: vLLM's default optimization level (O2) resolves cudagraph_mode to FULL_AND_PIECEWISE, whose decode_mode() is FULL, so `vllm serve google/diffusiongemma-26B-A4B-it` with no flags already runs every denoise/decode step under FULL cudagraphs.

I couldn't pin observed output corruption on this specifically: an E2E FULL-vs-PIECEWISE comparison was inconclusive, since DiffusionGemma's denoising has its own run-to-run variance of similar magnitude. The capture/replay test below reproduces the mechanism directly instead. So overall, this is a silent bug that likely has some unpredictable impact on output quality.

Fix: make `_causal_buf` int32 from construction and update it in place (`.copy_()` / direct assignment) instead of = from a bool source, so build() takes its dtype-matched fast path and never reallocates.

Tagging @LucasWilkinson, who authored the original FULL-cudagraph path in #45163.

AI-assisted, every line reviewed and tested by the submitter.

## Test Plan

`pytest tests/models/test_diffusion_gemma_cudagraph.py -v`

## Test Result

New regression test constructs a real DiffusionGemmaModelState and drives _causal_buf through an actual torch.cuda.CUDAGraph capture/replay cycle mirroring FlashAttention's cast. Fails on main: replay returns the stale `[0, 0, 0, 0]` instead of the updated `[1, 1, 1, 1]`. Passes with this patch. Run on H100 (SM90).

### CODE REVIEW

**Review by LucasWilkinson — COMMENTED** [review]
Good catch! thanks; can we remove the .to(...) call too inorder to avoid this footgun?

**Comment by fjosw** [comment]
Thanks @LucasWilkinson , I just pushed a new commit that removes the `.to(...)` call. `build()` now raises on non-int32 causal tensors instead, since the cast was out-of-place anyway and would silently detach graph replay from the caller's buffer. As far as I can tell, DiffusionGemma was the only in-tree model passing a tensor here.

**Comment by fjosw** [comment]
Given that this is a bug on the standard path in the official recipe, would you consider merging this @LucasWilkinson @DarkLight1337 @ywang96 @AndreasKaratzas @MatthewBonanni ?

**Comment by fjosw** [comment]
Hi @mmastrac, @LucasWilkinson, @njhill, @NickLucche I saw that you merged https://github.com/vllm-project/vllm/pull/57250 to add support for Jev-like structured outputs. Quick reminder that this PR fixes an issue where DiffusionGemma can use an incorrect attention mask during FULL CUDA graph replay, switching between causal and bidirectional attention incorrectly.

To my understanding, the example in https://github.com/vllm-project/vllm/pull/57250 uses `TRITON_ATTN` to which this bug does not apply, the default path should, however, still run into this bug.

The actual impact of a wrong attention mask is actually not super clear to me but the bug can impose a training-inference mismatch. I rebased the branch onto main and would recommend merging this simple fix.

**Review by mmastrac — APPROVED** [review]
This looks correct to me - will need a rebase but otherwise makes sense.

**Comment by mmastrac** [comment]
@fjosw LGTM. If you want to fix a related nit: `[num_seqs] bool` in `triton_unified_attention.py` should be `int32` (it was wrong before). 

@LucasWilkinson FWIW I think this should land as-is.

File: tests/models/test_diffusion_gemma_cudagraph.py
Comment (LucasWilkinson) [inline tests/models/test_diffusion_gemma_cudagraph.py:1]: nit: I think this test is a bit overly specific, not sure how much value it brings. I think we should probably just remove it

**Review by LucasWilkinson — APPROVED** [review]
LGTM thanks for the fix! sorry this slipped under my radar for so long! thanks @mmastrac for flagging
